### PR TITLE
BAAS-17702 add mutex to vm funcName to prevent data race

### PR DESCRIPTION
**Pull request by kpatel71716**
The issue has been present since at least May(it's a bit difficult going further back due to breaking baas changes). This might be caused by reusing the vm in an event loop and can make a ticket if we want to experiment with that route

### CODE REVIEW

**Comment by kpatel71716** [comment]
The data race occurs from concurrent access of the [vm.funcName](https://github.com/mongodb-forks/goja/blob/4373e9336f84e53b2de6f2e2ec7c73ba2487a1b9/vm.go#L204) variable and is occasionally reproducible with a function that is invoked from the event loop. Most data race alerts are from a read[ here](https://github.com/mongodb-forks/goja/blob/4373e9336f84e53b2de6f2e2ec7c73ba2487a1b9/func.go#L291) and a concurrent write [here](https://github.com/mongodb-forks/goja/blob/4373e9336f84e53b2de6f2e2ec7c73ba2487a1b9/func.go#L296) but this is reproducible in several other areas within the code. At first I thought this might be due to a larger issue but since this is code that we have changed ourselves in goja, my best guess is that we just never noticed this for a while until the alerts started firing from a potential change in charts e2e testing

**Review by LijieZhang1998 — APPROVED** [review]
LGTM 🚀

**Comment by kpatel71716** [comment]
> I think this is good enough and as we discussed online once we try to bump the version in baas we can check the perf impact of this change. Were you able to notice anything locally ?

Yep, have an eye on the benchmarks! I didn't notice anything locally so this _should_ be ok. I'll run the baas benchmarks locally and post those results before merging into baas

File: vm.go [hunk 1]
```diff
@@ -227,6 +228,19 @@ type instruction interface {
 	exec(*vm)
 }
 
+func (vm *vm) getFuncName() unistring.String {
+	vm.funcNameLock.Lock()
```
Comment (arahmanan) [inline vm.go:232]: It looks like we're using the write lock here and on line 234. You'll want this to be: 
```suggestion
	vm.funcNameLock.RLock()
```

File: vm.go [hunk 2]
```diff
@@ -227,6 +228,19 @@ type instruction interface {
...
+	s := vm.funcName
+	vm.funcNameLock.Unlock()
```
Comment (arahmanan) [inline vm.go:234]:
```suggestion
	vm.funcNameLock.RUnlock()
```

**Review by arahmanan — COMMENTED** [review]
Good find! Just a minor comment from my end.

File: vm.go (hunk 1 above)
Comment (kpatel71716) [reply vm.go]: 🤦 thanks for catching that!

**Pull request merged**